### ISSUE TITLE
mrconvert: enhanced 4D DICOM encoding formats are not fully supported

### ISSUE THREAD

**Issue by timrosenow** [MRtrix3/mrtrix3#2690]
Firstly, I apologise for the perhaps vague information in this bug report - DICOM encoding is far from my area of expertise!

Some 4D enhanced DICOM encoding formats are not supported by MRtrix3. Specifically, we have a 4D diffusion tensor image (from a Bruker scanner) that can be read and interpreted by MRtrix3 without issue. The functional group sequence encoding uses Item/SequenceDelimitationItems with undefined length. Upon upload to our DICOM server, this undergoes a re-encoding to sequences with explicit length. This is a valid DICOM format and can be viewed/opened by, for example, FSL. However, when opening with mrview, the images are corrupted (just noise), the axis dimensions are converted from [X,Y,Z,B] to [X,Y,1,Z*B], and the B-tables are truncated to just a single row.

**ORIGINAL IMAGE:**

```
trosenow@mrfellowpc:~/Downloads/tmp3$ mrinfo /data/research/chemobrain_mouse_43_chemobrain_mouse_43_chemobrain_E6_P1_EnIm1.dcm
mrinfo: [done] scanning DICOM folder "/data/rese..._chemobrain_E6_P1_EnIm1.dcm"
mrinfo: [100%] reading DICOM series "DTI_2seg_31dir b1/2000 0.2/0.4"
************************************************
Image name:          "chemobrain_mouse_43^^^^ (chemobrain_mouse_43) [MR] DTI_2seg_31dir b1/2000 0.2/0.4"
************************************************
 Dimensions:        100 x 52 x 44 x 67
 Voxel size:        0.2 x 0.2 x 0.4 x ?
 Data strides:      [ -1 -2 3 4 ]
 Format:            DICOM
 Data type:         unsigned 16 bit integer (little endian)
 Intensity scaling: offset = 0, multiplier = 33.769885284499999
 Transform:               0.9968    -0.07876    -0.01592      -9.056
                         0.07725      0.9938    -0.08014      -7.251
                         0.02213     0.07865      0.9967      -9.642
 FlipAngle:         90
 PixelBandwidth:    3000
 RepetitionTime:    3
 comments:          chemobrain_mouse_43^^^^ (chemobrain_mouse_43) [MR] DTI_2seg_31dir b1/2000 0.2/0.4
                    DOB: 20/06/2022
                    DOS: 20/06/2022 13:52:17
 dw_scheme:         0,0,0,68.374951296379706
 [67 entries]       0,0,0,68.374951296379706
                    ...
                    -0.47919169244646398,-0.87310715743715195,-0.089773122493923593,1027.969034393675
                    -0.47919169244646398,-0.87310715743715195,-0.089773122493923593,2025.5518261924087
```

**DOWNLOADED IMAGE**

```
trosenow@mrfellowpc:~/Downloads/tmp3$ mrinfo serverver/2.16.756.5.5.200.8323328.68896.1655774923.34-60001-1-1tl757z.dcm  
mrinfo: [done] scanning DICOM folder "serverver/...4923.34-60001-1-1tl757z.dcm"
mrinfo: [100%] reading DICOM series "DTI_2seg_31dir b1/2000 0.2/0.4"
************************************************
Image name:          "chemobrain_mouse_67^^^^ (chemobrain_mouse_67) [MR] DTI_2seg_31dir b1/2000 0.2/0.4"
************************************************
 Dimensions:        100 x 52 x 1 x 2948
 Voxel size:        0.2 x 0.2 x 0.4 x ?
 Data strides:      [ -2 -3 4 1 ]
 Format:            DICOM
 Data type:         unsigned 16 bit integer (little endian)
 Intensity scaling: offset = 0, multiplier = 41.234815597299999
 Transform:               0.9999    0.006102    -0.01069      -9.909
                       -0.009416      0.9385     -0.3452      -6.514
                        0.007925      0.3452      0.9385       5.889
 FlipAngle:         90
 PixelBandwidth:    3000
 RepetitionTime:    3
 comments:          chemobrain_mouse_67^^^^ (chemobrain_mouse_67) [MR] DTI_2seg_31dir b1/2000 0.2/0.4
                    study: chemobrain_mouse_67_chemobrain [ ORIGINAL PRIMARY DIFFUSION NONE ]
                    DOB: 21/06/2022
                    DOS: 21/06/2022 10:50:25
 dw_scheme:         -0.47919169244646398,-0.87310715743715195,-0.089773122493923593,2025.5518261924149 
```

**To Reproduce**

* Obtain a 4D enhanced dicom from a Bruker scanner (or maybe others?).
* Upload to any XNAT server (or other dcm4chee server).
* Download the re-encoded file
* View with mrview or convert with mrconvert.

The original scan and the corrupted server version can be found at this link:
https://filesender.aarnet.edu.au/?s=download&token=353ea827-b674-4493-b876-abbb006663d3 

Further discussion on the XNAT forum can be found here:
https://groups.google.com/g/xnat_discussion/c/8HvBDU1GT7c

**Platform/Environment/Version**

Please provide the following information:

-  OS: Ubuntu 22.04.2 LTS

-  *MRtrix3* version: == mrinfo 3.0.4-46-g8a192720 ==

---

See relevant DICOM standard document here: https://dicom.nema.org/medical/dicom/current/output/html/part05.html#sect_7.5

**Comment by stl-steve-moore** [comment]
I did some work with Tim, the person who posted. There is more detailed discussion in the XNAT link, but I can offer my comments here.

I believe that the mrtrix3 software does not properly parse DICOM images when Sequence elements use undefined lengths and delimiter elements rather that elements with a known length. The original data from the scanner has sequence elements with known lengths, and the software handles those. During the upload process (DICOM C-Store for example), the files are written to disk with sequences encoded with unknown lengths. I believe the encoding is correct, and that there is a parsing error in a library.

**Comment by Lestropie** [comment]
Thanks both for the very extensive explanation and providing exemplar data.

@jdtournier It'd likely take you less time to resolve this than it would I. Are you likely to find the time to look at it? If not I can give it a try.

**Comment by jdtournier** [comment]
OK, I'm on it.

**Comment by jdtournier** [comment]
Thanks for the report, problem confirmed at my end. Proposed fix in #2691...
